modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/utils.py

```diff
@@ -78,3 +82,61 @@ def _categories(self):  # noqa: GL08
     def _set_categories(self, categories):  # noqa: GL08
         self._lazy_categories = categories
         self._table = None
+
+
+def check_join_supported(join_type: str):
+    """
+    Check if join type is supported by HDK.
+
+    Parameters
+    ----------
+    join_type : str
+        Join type.
+
+    Returns
+    -------
+    None
+    """
+    if join_type not in ("inner", "left"):
+        raise NotImplementedError(
+            f"{join_type} join is not supported by the HDK engine"
+        )
+
+
+def arrow_to_pandas(at: pa.Table) -> pd.DataFrame:
+    """
+    Convert the specified arrow table to pandas.
+
+    Parameters
+    ----------
+    at : pyarrow.Table
+        The table to convert.
+
+    Returns
+    -------
+    pandas.DataFrame
+    """
+
+    def mapper(at):
+        if is_dictionary(at) and isinstance(at.value_type, ArrowIntervalType):
+            # The default mapper fails with TypeError: unhashable type: 'dict'
+            return _CategoricalDtypeMapper
+        return None
+
+    return at.to_pandas(types_mapper=mapper)
+
+
+class _CategoricalDtypeMapper:  # noqa: GL08
+    @staticmethod
+    def __from_arrow__(arr):  # noqa: GL08
+        values = []
+        categories = OrderedDict()
+        chunks = arr.chunks if isinstance(arr, pa.ChunkedArray) else (arr,)
+        for chunk in chunks:
+            assert isinstance(chunk, pa.DictionaryArray)
+            cat = chunk.dictionary.to_pandas()
+            values.extend([cat[v.index.as_py()] if v.is_valid else None for v in chunk])
+            categories.update((c, None) for c in cat)
+        return pd.Categorical(
+            values, dtype=pd.CategoricalDtype(categories, ordered=True)
+        )
```

Comment:
```suggestion
        pd_chunks = []
        for chunk in chunks:
            assert isinstance(chunk, pa.DictionaryArray)
            pd_chunks.append(chunk.indices.to_pandas().map(chunk.dictionary.to_pandas()))
        return pd.Categorical(pd.concat([pd_chunks]))
```
Should it work the same?
```python
>>> chunk
<pyarrow.lib.DictionaryArray object at 0x000001BB0B33D6D0>

-- dictionary:
  -- is_valid: all not null
  -- child 0 type: double
    [
      1.199,
      2
    ]
  -- child 1 type: double
    [
      2,
      4.4
    ]
-- indices:
  [
    null,
    null,
    0,
    0,
    1,
    1,
    null
  ]
>>> chunk.indices.to_pandas().map(chunk.dictionary.to_pandas())
0             NaN
1             NaN
2    (1.199, 2.0]
3    (1.199, 2.0]
4      (2.0, 4.4]
5      (2.0, 4.4]
6             NaN
dtype: interval
```

Reply:
With minor corrections, it works almost the same and looks much better, but it does not work properly in case of NaN values. There is a condition in the values mapping:
```python
[cat[v.index.as_py()] if v.is_valid else None
```

Reply:
could you please show an example where it doesn't work? In the suggestion I posted above, I try the conversion with the NaN values in the column and it works fine

Reply:
Sorry, I didn't clean the test env and my test failed. It works. But there is another issue - it does not preserve the categories order. Here is a simple reproducer:
```python
dtype = pandas.CategoricalDtype([pd.Interval(3, 4), pd.Interval(1, 2)], ordered=True)
pdf = pandas.DataFrame({"A": [pd.Interval(1, 2), pd.Interval(3, 4)]}, dtype=dtype)
at = pyarrow.Table.from_pandas(pdf)
mdf = arrow_to_pandas(at)
assert all(pdf.dtypes == mdf.dtypes)
```
For this reason OrderedDict is used there. I've applied your approach, but kept the dict.

Perhaps, we could ignore the order here. Anyway, with the HDK backend, in some cases, the order of categories is different from Pandas.